sensor/common/compliance/nodescan_handler_test.go

```diff
@@ -0,0 +1,161 @@
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+)
+
+func TestNodeScanHandler(t *testing.T) {
+	suite.Run(t, new(NodeScanHandlerTestSuite))
+}
+
+func fakeNodeScanV2(nodeName string) *storage.NodeScanV2 {
+	msg := &storage.NodeScanV2{
+		NodeId:   "",
+		NodeName: nodeName,
+		ScanTime: timestamp.TimestampNow(),
+		Components: &scannerV1.Components{
+			Namespace: "Testme OS",
+			RhelComponents: []*scannerV1.RHELComponent{
+				{
+					Name:      "vim-minimal",
+					Namespace: "rhel:8",
+					Version:   "2:7.4.629-6.el8.x86_64",
+					Arch:      "x86_64",
+					Module:    "FakeMod",
+					Cpes:      []string{"cpe:/a:redhat:enterprise_linux:8::baseos"},
+					AddedBy:   "FakeLayer",
+				},
+			},
+			LanguageComponents: nil,
+		},
+		Notes: nil,
+	}
+	return msg
+}
+
+type nodeScansProducerThatStops struct {
+	numMsgs   int
+	nodeScans chan *storage.NodeScanV2
+}
+
+type stoppable interface {
+	Stop(error)
+}
+
+func newNodeScansProducerThatStops(num int) *nodeScansProducerThatStops {
+	return &nodeScansProducerThatStops{
+		numMsgs:   num,
+		nodeScans: make(chan *storage.NodeScanV2),
+	}
+}
+
+func (nsp *nodeScansProducerThatStops) generator() <-chan *storage.NodeScanV2 {
+	return nsp.nodeScans
+}
+
+func (nsp *nodeScansProducerThatStops) start(s stoppable) {
+	go func(numMsgs int) {
+		defer close(nsp.nodeScans)
+		for i := 0; i < numMsgs; i++ {
+			nsp.nodeScans <- fakeNodeScanV2(fmt.Sprintf("Bobby_%d", i))
+			if i == 0 {
+				s.Stop(nil)
+			}
+		}
+	}(nsp.numMsgs)
+}
+
+type NodeScanHandlerTestSuite struct {
+	suite.Suite
+	mu                   *sync.Mutex
+	numReceivedAtCentral int
+	cancel               context.CancelFunc
+	ctx                  context.Context
+	toCentral            chan *central.MsgFromSensor
+}
+
+func (s *NodeScanHandlerTestSuite) SetupTest() {
+	s.mu = &sync.Mutex{}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.numReceivedAtCentral = 0
+	s.ctx, s.cancel = context.WithCancel(context.Background())
+	s.toCentral = make(chan *central.MsgFromSensor)
+}
+
+func (s *NodeScanHandlerTestSuite) TearDownTest() {
+
+}
+
+func (s *NodeScanHandlerTestSuite) startForwardingToCentral(c <-chan *central.MsgFromSensor) {
+	go func() {
+		defer close(s.toCentral)
+		for {
+			select {
+			case <-s.ctx.Done():
+				return
+			case msg, ok := <-c:
+				if !ok {
+					return
+				}
+				s.toCentral <- msg
+			}
+		}
+	}()
+}
+
+// startFakeCentral consumes all messages from s.toCentral and counts them
+func (s *NodeScanHandlerTestSuite) startFakeCentral() {
+	go func(ctx context.Context) {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-s.toCentral:
+				s.mu.Lock()
+				s.numReceivedAtCentral++
+				s.mu.Unlock()
+			}
+		}
+	}(s.ctx)
+}
+
+// TestStopHandler goal is to stop handler while there are still some messages to process
+// in the channel passed into NewNodeScanHandler.
+// We expect that premature stop of the handler produces no race condition or goroutine leak.
```

Comment:
How come no goroutine leaks if the one started by `nodeScansProducerThatStops.start()` should hang and leak after sending the first few items over its channel?

Reply:
Correct, it leaked. This should be fixed now (checked with https://github.com/uber-go/goleak, which I cannot include as the test logger used here leaks one goroutine)

Reply:
I played more with the leaks and the `goleak` from Uber. I was able to eliminate all leaks in the code (test & feat). I hope it is okay to add https://github.com/uber-go/goleak as a new dependency.